File: src/stargate_acceptor.erl
-module(stargate_acceptor).
-behaviour(gen_server).

-compile(export_all).

-include("global.hrl").

start_link(Params) -> gen_server:start(?MODULE, Params, []).

init({StargateArgs, ListenSocket, Id}) ->
    process_flag(trap_exit, true),
    {ok, _} = prim_inet:async_accept(ListenSocket, -1),
    {ok, #{params=> StargateArgs, listen_socket=> ListenSocket, id=> Id}}.


handle_info({inet_async, ListenSocket, _, {ok, ClientSocket}}, S=#{params:= Params}) ->
    prim_inet:async_accept(ListenSocket, -1),
    
    {ok, Pid} = vessel:start(Params),

    inet_db:register_socket(ClientSocket, inet_tcp),
    ok = gen_tcp:controlling_process(ClientSocket, Pid),

    Pid ! {pass_socket, ClientSocket},

    {noreply, S};

%TODO: Dont shutdown on all errors, see what is critical or not
handle_info({inet_async, _ListenSocket, _, Error}, S) ->
    ?PRINT({"Error in inet_async accept, shutting down", Error}),
    {stop, Error, S};

handle_info(_Message, S) -> {noreply, S}.

terminate(_Reason, S) -> 
    ?PRINT({"Terminated", _Reason, S}).


handle_cast(_Message, S) -> {noreply, S}.
handle_call(_Message, _From, S) -> {reply, ok, S}.
code_change(_OldVersion, S, _Extra) -> {ok, S}. 

File: lib/stargate_acceptor.ex
defmodule Stargate.Acceptor.Sup do
    def loop(config) do
        Process.flag(:trap_exit, true)
        loop_1(config)
    end

    def loop_1(config) do
        acceptors = Map.get(config,:acceptors,[])
        to_spawn = :erlang.system_info(:schedulers) - length(acceptors)
        acceptors = if to_spawn > 0 do
            pids = Enum.map(1..to_spawn, fn(_)-> :erlang.spawn_link(Stargate.Acceptor, :loop, [config]) end)
            acceptors ++ pids
        else acceptors end
        config = Map.put(config, :acceptors, acceptors)

        receive do
            {:EXIT, pid, r} ->
                IO.inspect {:stargate_acceptor_died, pid, r}
                loop_1(config)

            ukn -> 
                throw({__MODULE__, :ukn_msg, ukn})
        end
    end
end

defmodule Stargate.Acceptor do
    def loop(config) do
        {:ok, _} = :prim_inet.async_accept(config.listen_socket, -1)
        receive do
            {:inet_async, _ListenSocket, _, {:ok, csocket}} ->
                pid = :erlang.spawn(Stargate.Vessel, :loop, [config])

                :inet_db.register_socket(csocket, :inet_tcp)
                :ok = :gen_tcp.controlling_process(csocket, pid)
                send(pid, {:pass_socket, csocket})

                loop(config)

            {:inet_async, _, _, error} ->
                IO.inspect {:inet_async_error, error}
                loop(config)

            ukn -> 
                throw({__MODULE__, :ukn_msg, ukn})
        end
    end
end
